core/pkg/store/store.go

```diff
@@ -131,146 +210,180 @@ func (f *Store) String() (string, error) {
 }
 
 // GetAll returns a copy of the store's state (copy in order to be concurrency safe)
-func (f *Store) GetAll(_ context.Context) (map[string]model.Flag, model.Metadata, error) {
-	txn := f.db.Txn(false)
-
+func (s *Store) GetAll(ctx context.Context, selector *Selector, watcher chan Payload) (map[string]model.Flag, model.Metadata, error) {
+	txn := s.db.Txn(false)
 	flags := make(map[string]model.Flag)
-	it, err := txn.Get("flags", "id")
+	queryMeta := model.Metadata{}
+
+	var it memdb.ResultIterator
+	var err error
+	if selector != nil && !selector.isEmpty() {
+		queryMeta = selector.SelectorToMetadata()
+		indexId, constraints := selector.SelectorMapToQuery()
+		s.logger.Debug(fmt.Sprintf("getting all flags with query: %s, %v", indexId, constraints))
+		it, err = txn.Get("flags", indexId, constraints...)
+	} else {
+		// no selector, get all flags
+		it, err = txn.Get(flagsTable, idIndex)
+	}
 
 	if err != nil {
-		return flags, model.Metadata{}, err
+		s.logger.Error(fmt.Sprintf("query error: %v", err))
+		return flags, queryMeta, err
 	}
 
-	for obj := it.Next(); obj != nil; obj = it.Next() {
-		flag := obj.(model.Flag)
+	for raw := it.Next(); raw != nil; raw = it.Next() {
+		flag := raw.(model.Flag)
+		if existing, ok := flags[flag.Key]; ok {
+			if flag.Priority < existing.Priority {
+				s.logger.Debug(fmt.Sprintf("discarding duplicate flag %s from lower priority source %s in favor of flag from source %s", flag.Key, s.sources[flag.Priority], s.sources[existing.Priority]))
+				continue // we already have a higher priority flag
+			} else {
+				s.logger.Debug(fmt.Sprintf("overwriting duplicate flag %s from lower priority source %s in favor of flag from source %s", flag.Key, s.sources[existing.Priority], s.sources[flag.Priority]))
+			}
+		}
 		flags[flag.Key] = flag
 	}
 
-	return flags, f.getMetadata(), nil
+	if watcher != nil {
+
+		// a "one-time" watcher that will be notified of changes to the query
+		changes := it.WatchCh()
+
+		go func() {
```

Comment:
Can we have a separate function for watching? It can be modified to a for(ever) loop with the `WatchSet` provided by memdb.

```
    for {
        txn := db.Txn(false)
        ws  := memdb.NewWatchSet()

        iter, _ := txn.Get(...)
        ws.Add(iter.WatchCh())
        
        watcher <- ...
        
        if err := ws.WatchCtx(ctx); err != nil {
            close(watcher)
            return err                    // ctx cancelled or deadline
        }
        // loop restarts with a brand-new snapshot & watch set
    }
```

Reply:
This sounds much nicer than my optional recursion. Let me see...

Reply:
This is _much_ nicer. Excellent recommendation. Please see [this commit](https://github.com/open-feature/flagd/pull/1702/commits/cd87359ab339ed9208e0aeccdc5a5c160af4f3dd) for my changes (mostly related to this comment, but also some of your others).

Please note that to do add the `Watch` func you recommended here, I extracted some private functions to reduce some duplication (otherwise `GetAll` and `Watch` would have contained some trivial duplication). Despite that, I still think the ergonomics of your proposal are better, since in many cases we did a `GetAll` without a watching, and vice versa.

I also extracted the notifications as you suggested and added tests for that.